src/components/forms/TextInput.jsx

```diff
@@ -0,0 +1,143 @@
+      before,
+      ...props
+    },
+    ref
+  ) => {
+    // Refer to the following for an explanation on
+    // the throttling / debouncing approach:
+    // https://www.peterbe.com/plog/how-to-throttle-and-debounce-an-autocomplete-input-in-react
+    let onThrottledChangeThrottled;
+    let onThrottledChangeDebounced;
+
+    const assignThrottleFunctions = () => {
+      if (onThrottledChange) {
+        onThrottledChangeThrottled = throttle(throttleValue, onThrottledChange);
+        onThrottledChangeDebounced = debounce(throttleValue, onThrottledChange);
+      }
+    };
+
+    const inputOnChange = (e) => {
+      e.persist();
+      onChange(e);
+      if (onThrottledChange) {
+        throttleInput(
+          e.target.value,
+          onThrottledChangeThrottled,
+          onThrottledChangeDebounced
+        );
+      }
+    };
+
+    useEffect(assignThrottleFunctions, [onThrottledChange]);
+
+    const Element = showLabel ? 'label' : 'div';
+
+    return (
+      <>
+        <Element
+          className={cx('fdsTextInput-root', {
+            'fdsTextInput-root--horizontal': orientation === 'horizontal',
+          })}
+        >
+          {showLabel && label && (
+            <div className="fdsTextInput-label">
+              {label}
+              {showRequired && <span className="color--red">&nbsp;*</span>}
+            </div>
+          )}
+          <DecoratedInput after={after} before={before}>
+            <IconInput IconLeft={IconLeft} IconRight={IconRight}>
+              <input
+                {...props}
+                ref={ref}
+                aria-label={showLabel ? label : undefined}
+                onChange={inputOnChange}
+                type={type}
+                className={cx('fdsTextInput', {
+                  error: errorText,
+                })}
+              />
+            </IconInput>
+          </DecoratedInput>
+        </Element>
+        {errorText && <div className="color--red type--data">{errorText}</div>}
+      </>
+    );
+  }
+);
+
+TextInput.displayName = 'TextInput';
+
+TextInput.propTypes = {
+  errorText: PropTypes.string,
+
+  /** Label used for `label` element */
+  label: PropTypes.string.isReqiured,
+
+  /**
+   * FDS Icon _reference_
+   * `IconLeft={StarFilledIcon}`
+   */
+  IconLeft: PropTypes.func,
+
+  /**
+   * FDS Icon _reference_
+   * `IconRight={StarFilledIcon}`
+   */
+  IconRight: PropTypes.func,
+
+  /** Input element `type` attribute */
+  type: PropTypes.string,
+
+  /** Controls showing asterisk on label, signifying field is required */
+  required: PropTypes.bool,
+
+  /** Standard React onChange event */
+  onChange: PropTypes.func,
+
+  /** Custom onChange event that throttles / debounces. */
+  onThrottledChange: PropTypes.func,
+
+  /** String to place to the left of the input */
+  before: PropTypes.string,
+
+  /** String to place to the right of the input */
+  after: PropTypes.string,
+
+  orientation: PropTypes.oneOf(['vertical', 'horizontal']),
```

Comment:
Only thing left to address here - I suspect the `horizontal` orientation may more commonly need the label on the left of the input.

This also opens up a potential API change if we need to give users control over left/right/top

Reply:
@akdetrick added!